File: too_many_cooks/lib/src/tools/plan_tool.dart
/// Plan tool - agent plan management.
library;

import 'package:dart_logging/dart_logging.dart';
import 'package:dart_node_mcp/dart_node_mcp.dart';
import 'package:nadz/nadz.dart';
import 'package:too_many_cooks/src/notifications.dart';
import 'package:too_many_cooks/src/types.dart';
import 'package:too_many_cooks_data/too_many_cooks_data.dart';

/// Input schema for plan tool.
const planInputSchema = <String, Object?>{
  'type': 'object',
  'properties': {
    'action': {
      'type': 'string',
      'enum': ['update', 'get', 'list'],
      'description': 'Plan action to perform',
    },
    'goal': {
      'type': 'string',
      'maxLength': 100,
      'description': 'Your goal (for update). MUST be 100 chars or less.',
    },
    'current_task': {
      'type': 'string',
      'maxLength': 100,
      'description':
          'What you are doing now (for update). '
          'MUST be 100 chars or less.',
    },
  },
  'required': ['action'],
};

/// Tool config for plan.
const planToolConfig = (
  title: 'Plan',
  description:
      'Manage your plan. You must register first (except list). '
      'REQUIRED: action (update|get|list). '
      'For update: goal, current_task. '
      'Example: {"action":"update","goal":"Fix bug",'
      ' "current_task":"Reading code"}',
  inputSchema: planInputSchema,
  outputSchema: null,
  annotations: null,
);

/// Create plan tool handler.
ToolCallback createPlanHandler(
  TooManyCooksDb db,
  NotificationEmitter emitter,
  Logger logger,
  SessionGetter getSession,
) => (args, meta) async {
  final actionArg = args['action'];
  if (actionArg == null || actionArg is! String) {
    return (
      content: <Object>[
        textContent('{"error":"missing_parameter: action is required"}'),
      ],
      isError: true,
    );
  }
  final action = actionArg;
  final log = logger.child({'tool': 'plan', 'action': action});

  // list doesn't need auth
  if (action == 'list') return _list(db);

  // Hidden agent_key override for multi-agent integration testing
  final keyOverride = args['agent_key'] as String?;
  final String agentName;
  final String agentKey;
  if (keyOverride != null) {
    agentKey = keyOverride;
    switch (db.lookupByKey(keyOverride)) {
      case Success(:final value):
        agentName = value;
      case Error(:final error):
        return _errorResult(error);
    }
  } else {
    final session = getSession();
    if (session == null) {
      return (
        content: <Object>[
          textContent('{"error":"not_registered: call register first"}'),
        ],
        isError: true,
      );
    }
    agentName = session.agentName;
    agentKey = session.agentKey;
  }

  return switch (action) {
    'update' => _update(
      db,
      emitter,
      log,
      agentName,
      agentKey,
      args['goal'] as String?,
      args['current_task'] as String?,
    ),
    'get' => _get(db, agentName),
    _ => (
      content: <Object>[textContent('{"error":"Unknown action: $action"}')],
      isError: true,
    ),
  };
};

CallToolResult _update(
  TooManyCooksDb db,
  NotificationEmitter emitter,
  Logger log,
  String agentName,
  String agentKey,
  String? goal,
  String? currentTask,
) {
  if (goal == null || currentTask == null) {
    return (
      content: <Object>[
        textContent('{"error":"update requires goal, current_task"}'),
      ],
      isError: true,
    );
  }
  return switch (db.updatePlan(agentName, agentKey, goal, currentTask)) {
    Success() => () {
      emitter.emit(eventPlanUpdated, {
        'agent_name': agentName,
        'goal': goal,
        'current_task': currentTask,
      });
      log.info('Plan updated for $agentName: $currentTask');
      return (
        content: <Object>[textContent('{"updated":true}')],
        isError: false,
      );
    }(),
    Error(:final error) => _errorResult(error),
  };
}

CallToolResult _get(TooManyCooksDb db, String agentName) =>
    switch (db.getPlan(agentName)) {
      Success(:final value) when value == null => (
        content: <Object>[textContent('{"plan":null}')],
        isError: false,
      ),
      Success(:final value) => (
        content: <Object>[textContent('{"plan":${agentPlanToJson(value!)}}')],
        isError: false,
      ),
      Error(:final error) => _errorResult(error),
    };

CallToolResult _list(TooManyCooksDb db) => switch (db.listPlans()) {
  Success(:final value) => (
    content: <Object>[
      textContent('{"plans":[${value.map(agentPlanToJson).join(',')}]}'),
    ],
    isError: false,
  ),
  Error(:final error) => _errorResult(error),
};

CallToolResult _errorResult(DbError e) =>
    (content: <Object>[textContent(dbErrorToJson(e))], isError: true);

File: too-many-cooks/packages/core/src/tools/plan_tool.ts
/// Plan tool - agent plan management.

import type { Logger } from "../logger.js";
import type { NotificationEmitter } from "../notifications.js";
import { EVENT_PLAN_UPDATED } from "../notifications.js";
import type { TooManyCooksDb } from "../db-interface.js";
import type { Result } from "../result.js";
import type { AgentPlan, DbError } from "../types.js";
import { agentPlanToJson } from "../types.js";
import {
  type CallToolResult,
  type SessionGetter,
  type ToolCallback,
  textContent,
} from "../mcp-types.js";
import { type IdentityResult, errorContent, makeErrorResult, resolveIdentity } from "./tool_utils.js";

/** Input schema for plan tool. */
export const PLAN_INPUT_SCHEMA: {
  readonly type: "object";
  readonly properties: Record<string, unknown>;
  readonly required: readonly string[];
} = {
  type: "object",
  properties: {
    action: {
      type: "string",
      enum: ["update", "get", "list"],
      description: "Plan action to perform",
    },
    goal: {
      type: "string",
      maxLength: 100,
      description: "Your goal (for update). MUST be 100 chars or less.",
    },
    current_task: {
      type: "string",
      maxLength: 100,
      description:
        "What you are doing now (for update). MUST be 100 chars or less.",
    },
  },
  required: ["action"],
} as const;

/** Tool config for plan. */
export const PLAN_TOOL_CONFIG: {
  readonly title: string;
  readonly description: string;
  readonly inputSchema: typeof PLAN_INPUT_SCHEMA;
  readonly outputSchema: null;
  readonly annotations: null;
} = {
  title: "Plan",
  description:
    "Manage your plan. You must register first (except list). " +
    "REQUIRED: action (update|get|list). " +
    "For update: goal, current_task. " +
    'Example: {"action":"update","goal":"Fix bug",' +
    ' "current_task":"Reading code"}',
  inputSchema: PLAN_INPUT_SCHEMA,
  outputSchema: null,
  annotations: null,
} as const;

/** Create plan tool handler. */
export const createPlanHandler: (
  db: TooManyCooksDb,
  emitter: NotificationEmitter,
  logger: Logger,
  getSession: SessionGetter,
) => ToolCallback = (
  db: TooManyCooksDb,
  emitter: NotificationEmitter,
  logger: Logger,
  getSession: SessionGetter,
): ToolCallback =>
  {return async (args: Record<string, unknown>): Promise<CallToolResult> => {
    const actionArg: unknown = args.action;
    if (typeof actionArg !== "string") {
      return errorContent("missing_parameter: action is required");
    }
    const action: string = actionArg;
    const log: Logger = logger.child({ tool: "plan", action });

    if (action === "list") {return await handleList(db);}

    const identity: IdentityResult = await resolveIdentity(db, args, getSession);
    if (identity.isError) {return identity.result;}
    const { agentName, agentKey }: { agentName: string; agentKey: string } = identity;

    switch (action) {
      case "update":
        return await handleUpdate(
          db,
          emitter,
          log,
          agentName,
          agentKey,
          typeof args.goal === "string" ? args.goal : null,
          typeof args.current_task === "string"
            ? args.current_task
            : null,
        );
      case "get":
        return await handleGet(db, agentName);
      default:
        return {
          content: [
            textContent(
              JSON.stringify({ error: `Unknown action: ${action}` }),
            ),
          ],
          isError: true,
        };
    }
  }};

// ---------------------------------------------------------------------------
// Update
// ---------------------------------------------------------------------------

const handleUpdate: (
  db: TooManyCooksDb,
  emitter: NotificationEmitter,
  log: Logger,
  agentName: string,
  agentKey: string,
  goal: string | null,
  currentTask: string | null,
) => Promise<CallToolResult> = async (
  db: TooManyCooksDb,
  emitter: NotificationEmitter,
  log: Logger,
  agentName: string,
  agentKey: string,
  goal: string | null,
  currentTask: string | null,
): Promise<CallToolResult> => {
  if (goal === null || currentTask === null) {
    return errorContent("update requires goal, current_task");
  }
  const result: Result<void, DbError> = await db.updatePlan(agentName, agentKey, goal, currentTask);
  if (!result.ok) {return makeErrorResult(result.error);}
  emitter.emit(EVENT_PLAN_UPDATED, {
    agent_name: agentName,
    goal,
    current_task: currentTask,
  });
  log.info(`Plan updated for ${agentName}: ${currentTask}`);
  return {
    content: [textContent(JSON.stringify({ updated: true }))],
    isError: false,
  };
};

// ---------------------------------------------------------------------------
// Get
// ---------------------------------------------------------------------------

const handleGet: (
  db: TooManyCooksDb,
  agentName: string,
) => Promise<CallToolResult> = async (
  db: TooManyCooksDb,
  agentName: string,
): Promise<CallToolResult> => {
  const result: Result<AgentPlan | null, DbError> = await db.getPlan(agentName);
  if (!result.ok) {return makeErrorResult(result.error);}
  if (result.value !== null) {
    return {
      content: [
        textContent(
          JSON.stringify({ plan: agentPlanToJson(result.value) }),
        ),
      ],
      isError: false,
    };
  }
  return {
    content: [textContent(JSON.stringify({ plan: null }))],
    isError: false,
  };
};

// ---------------------------------------------------------------------------
// List
// ---------------------------------------------------------------------------

const handleList: (db: TooManyCooksDb) => Promise<CallToolResult> = async (db: TooManyCooksDb): Promise<CallToolResult> => {
  const result: Result<readonly AgentPlan[], DbError> = await db.listPlans();
  if (!result.ok) {return makeErrorResult(result.error);}
  return {
    content: [
      textContent(
        JSON.stringify({ plans: result.value.map(agentPlanToJson) }),
      ),
    ],
    isError: false,
  };
};
